This is a backport from gcc bugzilla

http://gcc.gnu.org/bugzilla/show_bug.cgi?id=55383

fixes wrong warning behavior which has no workaround
other than disabling the warning

Upstream-Status: Backport

Signed-off-by: Khem Raj <raj.khem@gmail.com>

Index: gcc-4.7.2/gcc/c-typeck.c
===================================================================
--- gcc-4.7.2.orig/gcc/c-typeck.c	2012-09-13 10:04:44.000000000 -0700
+++ gcc-4.7.2/gcc/c-typeck.c	2013-03-10 23:09:17.610899536 -0700
@@ -4557,7 +4557,7 @@
     /* There are qualifiers present in IN_OTYPE that are not present
        in IN_TYPE.  */
     warning_at (loc, OPT_Wcast_qual,
-		"cast discards %q#v qualifier from pointer target type",
+		"cast discards %qv qualifier from pointer target type",
 		discarded);
 
   if (added || discarded)
Index: gcc-4.7.2/gcc/testsuite/c-c++-common/Wcast-qual-1.c
===================================================================
--- gcc-4.7.2.orig/gcc/testsuite/c-c++-common/Wcast-qual-1.c	2011-04-07 14:47:38.000000000 -0700
+++ gcc-4.7.2/gcc/testsuite/c-c++-common/Wcast-qual-1.c	2013-03-10 23:15:45.610894589 -0700
@@ -85,11 +85,11 @@
 void
 f4 (void * const **bar)
 {
-  const void ***p9 = (const void ***) bar; /* { dg-warning "cast" } */
+  const void ***p9 = (const void ***) bar; /* { dg-warning "cast discards .const. qualifier" } */
   void * const **p11 = (void * const **) bar;
-  void ** const *p13 = (void ** const *) bar; /* { dg-warning "cast" } */
+  void ** const *p13 = (void ** const *) bar; /* { dg-warning "cast discards .const. qualifier" } */
   const void * const **p15 = (const void * const **) bar; /* { dg-warning "cast" } */
-  const void ** const *p17 = (const void ** const *) bar; /* { dg-warning "cast" } */
+  const void ** const *p17 = (const void ** const *) bar; /* { dg-warning "cast discards .const. qualifier" } */
   void * const * const * p19 = (void * const * const *) bar;
   const void * const * const *p21 = (const void * const * const *) bar;
 }
